ui/common/src/commonMain/kotlin/social/androiddev/common/composables/text/MastodonTextField.kt

```diff
@@ -0,0 +1,301 @@
+
+    @Composable
+    override fun placeholderColor(enabled: Boolean): State<Color> {
+        return rememberUpdatedState(if (enabled) placeholderColor else disabledPlaceholderColor)
+    }
+
+    @Composable
+    override fun labelColor(
+        enabled: Boolean,
+        error: Boolean,
+        interactionSource: InteractionSource
+    ): State<Color> {
+        val focused by interactionSource.collectIsFocusedAsState()
+
+        val targetValue = when {
+            !enabled -> disabledLabelColor
+            error -> errorLabelColor
+            focused -> focusedLabelColor
+            else -> unfocusedLabelColor
+        }
+        return rememberUpdatedState(targetValue)
+    }
+
+    @Composable
+    override fun textColor(enabled: Boolean): State<Color> {
+        return rememberUpdatedState(if (enabled) textColor else disabledTextColor)
+    }
+
+    @Composable
+    override fun cursorColor(isError: Boolean): State<Color> {
+        return rememberUpdatedState(if (isError) errorCursorColor else cursorColor)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as DefaultTextFieldColors
+
+        if (textColor != other.textColor) return false
+        if (disabledTextColor != other.disabledTextColor) return false
+        if (cursorColor != other.cursorColor) return false
+        if (errorCursorColor != other.errorCursorColor) return false
+        if (focusedIndicatorColor != other.focusedIndicatorColor) return false
+        if (unfocusedIndicatorColor != other.unfocusedIndicatorColor) return false
+        if (errorIndicatorColor != other.errorIndicatorColor) return false
+        if (disabledIndicatorColor != other.disabledIndicatorColor) return false
+        if (leadingIconColor != other.leadingIconColor) return false
+        if (disabledLeadingIconColor != other.disabledLeadingIconColor) return false
+        if (errorLeadingIconColor != other.errorLeadingIconColor) return false
+        if (trailingIconColor != other.trailingIconColor) return false
+        if (disabledTrailingIconColor != other.disabledTrailingIconColor) return false
+        if (errorTrailingIconColor != other.errorTrailingIconColor) return false
+        if (backgroundColor != other.backgroundColor) return false
+        if (focusedLabelColor != other.focusedLabelColor) return false
+        if (unfocusedLabelColor != other.unfocusedLabelColor) return false
+        if (disabledLabelColor != other.disabledLabelColor) return false
+        if (errorLabelColor != other.errorLabelColor) return false
+        if (placeholderColor != other.placeholderColor) return false
+        if (disabledPlaceholderColor != other.disabledPlaceholderColor) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = textColor.hashCode()
+        result = 31 * result + disabledTextColor.hashCode()
```

Comment:
This is due to not having access to something like Objects.hash?

Reply:
Not sure - I mimicked what the built-in `TextField` composable does under the hood. My guess is
`androidx.compose.ui.graphics.Compose` is not a data class so in order to fulfill the `@Immutable` contract for `DefaultTextFieldColors` we need manually calculate the hashCode for all color properties